proposals/extensions_v2.md

```diff
@@ -0,0 +1,375 @@
+# Modern Extensions
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+Many thanks to those who helped with this proposal.  Esp. @jnm2!
+
+## Summary
+[summary]: #summary
+
+Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.
+
+This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).
+
+Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with classic extension methods, and then potential enhancements on top of this kernel to make certain common designs more succinct and pleasant.  This does not preclude the feature shipping with any or all of these enhancements at launch.  It simply allows the design to be broken into separable concerns, with a clearer ordering of dependencies.
+
+A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:
+
+Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.
+
+```c#
+extension E
+{
+    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
+    public int M<X>(...) for SomeType<X> val { }
+
+    // Property form:
+    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
+    public int Count<X> for SomeType<X> { get { ... } }
```

Comment:
It's confusing me that many of these don't have a parameter name, since the core proposal isn't making the parameter name optional on instance extensions.

Reply:
ok.  adding.